Storage/SQL/SQLiteRemoteClientsAndTabs.swift

```diff
@@ -211,49 +162,40 @@ open class SQLiteRemoteClientsAndTabs: RemoteClientsAndTabs {
                 succeeded += 1
             }
 
-            deferred.fill(Maybe(success: succeeded))
-            return true
+            return succeeded
         }
-
-        return deferred
     }
 
     open func insertOrUpdateClient(_ client: RemoteClient) -> Deferred<Maybe<Int>> {
         return insertOrUpdateClients([client])
     }
 
     open func deleteClient(guid: GUID) -> Success {
-        let deferred = Success(defaultQueue: DispatchQueue.main)
-
         let deleteTabsQuery = "DELETE FROM \(TableTabs) WHERE client_guid = ?"
         let deleteClientQuery = "DELETE FROM \(TableClients) WHERE guid = ?"
         let deleteArgs: Args = [guid]
 
-        var err: NSError?
+        return db.transaction { connection -> Void in
+            var err: NSError? = nil
 
-        _ = db.transaction(&err) { connection, _ in
-            var success = true
-            if let _ = connection.executeChange(deleteClientQuery, withArgs: deleteArgs) {
+            do {
+                try connection.executeChange(deleteClientQuery, withArgs: deleteArgs)
+            } catch let error as NSError {
                 log.warning("Deleting client failed.")
-                success = false
+                err = error
             }
 
-            // Delete any existing tabs.
-            if let _ = connection.executeChange(deleteTabsQuery, withArgs: deleteArgs) {
+            do {
+                try connection.executeChange(deleteTabsQuery, withArgs: deleteArgs)
+            } catch let error as NSError {
                 log.warning("Deleting client tabs failed.")
-
-                success = false
+                err = error
             }
 
-            if success {
-                deferred.fill(Maybe(success: ()))
-            } else {
-                deferred.fill(Maybe(failure: DatabaseError(err: err)))
+            if let err = err {
+                throw err
```

Comment:
If you throw here, the transaction is rolled back, so you might as well throw immediately, and get rid of all that `err` stuff!

But in any case, `TableTabs` is `ON DELETE CASCADE`, so if foreign key support is on — which it should be — we shouldn't even need to delete the tabs here.

Simplify!